lib/shrine/plugins/derivates.rb

```diff
@@ -0,0 +1,151 @@
+        uploader.opts[:versions_fallback_to_original] = opts.fetch(:fallback_to_original, uploader.opts.fetch(:versions_fallback_to_original, true))
+
+        unless [:separate, :original].include? uploader.opts[:derivates_attribute]
+          raise Error, "`#{uploader.opts[:derivates_attribute] || "nil"}` is an invalid " \
+            "option for `data_attribute` in Derivates Plugin. Valid options are: " \
+            "[:separate, :original]"
+        end
+      end
+
+      module ClassMethods
+        def version_fallbacks
+          opts[:version_fallbacks]
+        end
+      end
+
+      module AttachmentMethods
+        def initialize(name, **options)
+          super
+
+          module_eval <<-RUBY, __FILE__, __LINE__ + 1
+            def #{name}_derivates
+              #{name}_attacher.get_derivates
+            end
+          RUBY
+        end
+      end
+
+      module AttacherMethods
+        def changed?
+          instance_variable_defined?(:@old) || instance_variable_defined?(:@derivates_queue)
+        end
+
+        def finalize
+          super
+          update_derivates
+        end
+
+        def add_derivate(key, io, store = :store, nested_list = nil)
+          @derivates_queue ||= {}
+          if store == :store
+            @derivates_queue.merge! key.to_sym => store!(io)
+          else
+            derivate_attacher = shrine_class::Attacher.new(context[:record], context[:name], cache: @cache.storage_key, store: store)
+            @derivates_queue.merge! key.to_sym => derivate_attacher.store!(io)
+          end
+        end
+
+        def update_derivates
+          if instance_variable_defined?(:@derivates_queue)
+            _set_derivates(@derivates_queue)
+            remove_instance_variable(:@derivates_queue)
+          end
+        end
+
+        def get
+          data = read
+          if data
+            if derivates_attribute == :original
+              data = JSON.parse(data)
+              data.delete "derivates" if data.has_key? "derivates"
+            end
+            uploaded_file(data)
+          end
+        end
+
+        def get_derivates
+          data = read_derivates
+          data = JSON.parse(data) if data
+          convert_to_uploaded_file(data["derivates"])
+        end
+
+        private
+
+        def convert_to_uploaded_file(object)
+          return nil if object.nil?
+
+          if object.is_a?(Hash) && object.values.none? { |value| value.is_a?(String) }
+            object.inject({}) do |result, (name, value)|
+              result.merge!(name.to_sym => convert_to_uploaded_file(value))
+            end
+          elsif object.is_a?(Array)
+            object.map { |value| convert_to_uploaded_file(value) }
+          else
+            uploaded_file(object)
+          end
+        end
+
+        def derivates_attribute
+          shrine_class.opts[:derivates_attribute]
+        end
+
+        def _set_derivates(uploaded_file)
```

Comment:
This method will always overwrite derivates the list. As discussed, we ideally want a method that can add/remove attachments to the current list.